src/test/isolation2/sql/frozen_insert_crash.sql

```diff
@@ -199,6 +199,21 @@ $$ LANGUAGE plpgsql;
 0U: select insert_bm_lov_res();
 0U: select * from bm_lov_res;
 
+1: drop extension pageinspect;
+
+-- Test for aoseg: suspend the insert into aoseg table before we mark the row frozen.
+-- Another session should still be able to choose a different segno.
+1: create table tab_aoseg(a int) with (appendoptimized=true, orientation=row);
+1: select gp_inject_fault('insert_aoseg_before_freeze', 'suspend', dbid) from gp_segment_configuration where role = 'p' and content = 0;
```

Comment:
Why did we remove additional begin's and end's from here?

Reply:
Because with them the test hangs. Moreover, this test does not check anything even on 7X! (see tracker for details)

Reply:
With begin on 6X test hangs on command
```sql
1>: insert into tab_aoseg select * from generate_series(1,10);
```
after command
```sql
3: select gp_inject_fault('insert_aoseg_before_freeze', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
```
with stack
```
libc.so.6!semop() (/usr/src/debug/glibc-2.17-c758a686/sysdeps/unix/syscall-template.S:81)
PGSemaphoreLock(PGSemaphore sema, bool interruptOK) (src/backend/port/sysv_sema.c:422)
ProcSleep(LOCALLOCK * locallock, LockMethod lockMethodTable) (src/backend/storage/lmgr/proc.c:1351)
WaitOnLock(LOCALLOCK * locallock, ResourceOwner owner) (src/backend/storage/lmgr/lock.c:1867)
LockAcquireExtended(const LOCKTAG * locktag, LOCKMODE lockmode, bool sessionLock, bool dontWait, bool reportMemoryError, LOCALLOCK ** locallockp) (src/backend/storage/lmgr/lock.c:1155)
LockRelationOid(Oid relid, LOCKMODE lockmode) (src/backend/storage/lmgr/lmgr.c:105)
try_relation_open(Oid relationId, LOCKMODE lockmode, bool noWait) (src/backend/access/heap/heapam.c:1126)
analyze_rel_internal(Oid relid, VacuumStmt * vacstmt, bool in_outer_xact, BufferAccessStrategy bstrategy) (src/backend/commands/analyze.c:273)
analyze_rel(Oid relid, VacuumStmt * vacstmt, bool in_outer_xact, BufferAccessStrategy bstrategy) (src/backend/commands/analyze.c:227)
vacuum(VacuumStmt * vacstmt, Oid relid, bool do_toast, BufferAccessStrategy bstrategy, bool for_wraparound, bool isTopLevel) (src/backend/commands/vacuum.c:388)
autostats_issue_analyze(Oid relationOid) (src/backend/postmaster/autostats.c:80)
auto_stats(AutoStatsCmdType cmdType, Oid relationOid, uint64 ntuples, bool inFunction) (src/backend/postmaster/autostats.c:332)
ProcessQuery(Portal portal, PlannedStmt * stmt, const char * sourceText, ParamListInfo params, DestReceiver * dest, char * completionTag) (src/backend/tcop/pquery.c:343)
PortalRunMulti(Portal portal, bool isTopLevel, DestReceiver * dest, DestReceiver * altdest, char * completionTag) (src/backend/tcop/pquery.c:1490)
PortalRun(Portal portal, int64 count, bool isTopLevel, DestReceiver * dest, DestReceiver * altdest, char * completionTag) (src/backend/tcop/pquery.c:1035)
exec_simple_query(const char * query_string) (src/backend/tcop/postgres.c:1852)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (src/backend/tcop/postgres.c:5332)
BackendRun(Port * port) (src/backend/postmaster/postmaster.c:4843)
BackendStartup(Port * port) (src/backend/postmaster/postmaster.c:4507)
ServerLoop() (src/backend/postmaster/postmaster.c:1957)
```
On 7X test does not hang, because commit 6776142a07afb4c28961f27059d800196902f5f1 added VACOPT_SKIP_LOCKED flag to skip if cannot get lock.

Reply:
As I already [noticed](https://github.com/arenadata/gpdb/pull/585#issuecomment-1713970889) this test does not test anything even on 7X.

Reply:
How this explains removal of begin and end? How this answers a question about vacuum?

Reply:
> This test contains no vacuum.

> How this answers a question about vacuum?

The first insertion into the table generates its analysis, which is done by a vacuum.

Reply:
> How this explains removal of begin and end?

Vacuum in 6X does not have a flag to skip blocked ones, so simultaneous inserts into the same table cannot begin until the transaction with the first insert completes.

Reply:
> Vacuum in 6X does not have a flag to skip blocked ones, so simultaneous inserts into the same table cannot begin until the transaction with the first insert completes.

Good point. Have you tried not to remove test lines, but to move upper ending of second session?

Reply:
No, may be I should try gp_autostats_mode = none ?

Reply:
I think a simple moving is the best option and not breaks the _idea_ of original test.

Reply:
such diff
```diff
diff --git a/src/test/isolation2/sql/frozen_insert_crash.sql b/src/test/isolation2/sql/frozen_insert_crash.sql
index 93c58b59e8..32b24943d7 100644
--- a/src/test/isolation2/sql/frozen_insert_crash.sql
+++ b/src/test/isolation2/sql/frozen_insert_crash.sql
@@ -205,13 +205,16 @@ $$ LANGUAGE plpgsql;
 -- Another session should still be able to choose a different segno.
 1: create table tab_aoseg(a int) with (appendoptimized=true, orientation=row);
 1: select gp_inject_fault('insert_aoseg_before_freeze', 'suspend', dbid) from gp_segment_configuration where role = 'p' and content = 0;
+1: begin;
 1>: insert into tab_aoseg select * from generate_series(1,10);
 -- wait until the aoseg record is inserted but not yet frozen
 2: select gp_wait_until_triggered_fault('insert_aoseg_before_freeze', 1, dbid) from gp_segment_configuration where role = 'p' and content = 0;
-2>: insert into tab_aoseg select * from generate_series(1,10);
+2: begin;
+2: insert into tab_aoseg select * from generate_series(1,10);
+2: end;
 3: select gp_inject_fault('insert_aoseg_before_freeze', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
 1<:
-2<:
+1: end;
 0U: select segno, eof from gp_toolkit.__gp_aoseg('tab_aoseg');
 
 -- validate that we've actually tested desired scan method
```
is working
or you mean something other?

Reply:
this diff
```diff
diff --git a/src/test/isolation2/sql/frozen_insert_crash.sql b/src/test/isolation2/sql/frozen_insert_crash.sql
index 93c58b59e8..57667575f0 100644
--- a/src/test/isolation2/sql/frozen_insert_crash.sql
+++ b/src/test/isolation2/sql/frozen_insert_crash.sql
@@ -205,13 +205,17 @@ $$ LANGUAGE plpgsql;
 -- Another session should still be able to choose a different segno.
 1: create table tab_aoseg(a int) with (appendoptimized=true, orientation=row);
 1: select gp_inject_fault('insert_aoseg_before_freeze', 'suspend', dbid) from gp_segment_configuration where role = 'p' and content = 0;
+1: begin;
 1>: insert into tab_aoseg select * from generate_series(1,10);
 -- wait until the aoseg record is inserted but not yet frozen
 2: select gp_wait_until_triggered_fault('insert_aoseg_before_freeze', 1, dbid) from gp_segment_configuration where role = 'p' and content = 0;
+2: begin;
 2>: insert into tab_aoseg select * from generate_series(1,10);
 3: select gp_inject_fault('insert_aoseg_before_freeze', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
-1<:
 2<:
+2: end;
+1<:
+1: end;
 0U: select segno, eof from gp_toolkit.__gp_aoseg('tab_aoseg');
 
 -- validate that we've actually tested desired scan method
```
is working too

Reply:
but this diff
```diff
diff --git a/src/test/isolation2/sql/frozen_insert_crash.sql b/src/test/isolation2/sql/frozen_insert_crash.sql
index 93c58b59e8..f6704ca3d3 100644
--- a/src/test/isolation2/sql/frozen_insert_crash.sql
+++ b/src/test/isolation2/sql/frozen_insert_crash.sql
@@ -205,13 +205,17 @@ $$ LANGUAGE plpgsql;
 -- Another session should still be able to choose a different segno.
 1: create table tab_aoseg(a int) with (appendoptimized=true, orientation=row);
 1: select gp_inject_fault('insert_aoseg_before_freeze', 'suspend', dbid) from gp_segment_configuration where role = 'p' and content = 0;
+1: begin;
 1>: insert into tab_aoseg select * from generate_series(1,10);
 -- wait until the aoseg record is inserted but not yet frozen
 2: select gp_wait_until_triggered_fault('insert_aoseg_before_freeze', 1, dbid) from gp_segment_configuration where role = 'p' and content = 0;
+2: begin;
 2>: insert into tab_aoseg select * from generate_series(1,10);
 3: select gp_inject_fault('insert_aoseg_before_freeze', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
-1<:
 2<:
+1<:
+2: end;
+1: end;
 0U: select segno, eof from gp_toolkit.__gp_aoseg('tab_aoseg');
 
 -- validate that we've actually tested desired scan method
```
already is not working

Reply:
> this diff

I think this diff is closer to original commit.

Reply:
Should I repush last two commits?

Reply:
If we want to squash all of this, then I think it's not necessary and you can push a separate commit with a fix.

Reply:
I think we want use all 5 separate commits.

Reply:
I repushed last two commits